spec/compiler/semantic/call_error_spec.cr

```diff
@@ -0,0 +1,207 @@
+    assert_error %(
+      def foo(x)
+      end
+
+      foo
+      ),
+      "wrong number of arguments for 'foo' (given 0, expected 1)"
+  end
+
+  it "says wrong number of arguments even if other overloads don't match by block" do
+    assert_error %(
+      def foo(x)
+      end
+
+      def foo(x, y)
+        yield
+      end
+
+      foo
+      ),
+      "wrong number of arguments for 'foo' (given 0, expected 1)"
+  end
+
+  it "says not expected to be invoked with a block" do
+    assert_error %(
+      def foo
+      end
+
+      foo {}
+      ),
+      "'foo' is not expected to be invoked with a block, but a block was given"
+  end
+
+  it "says expected to be invoked with a block" do
+    assert_error %(
+      def foo
+        yield
+      end
+
+      foo
+      ),
+      "'foo' is expected to be invoked with a block, but no block was given"
+  end
+
+  it "says missing named argument" do
+    assert_error %(
+      def foo(*, x)
+      end
+
+      foo
+      ),
+      "missing argument: x"
+  end
+
+  it "says missing named arguments" do
+    assert_error %(
+      def foo(*, x, y)
+      end
+
+      foo
+      ),
+      "missing arguments: x, y"
+  end
+
+  it "says no parameter named" do
+    assert_error %(
+      def foo
+      end
+
+      foo(x: 1)
+      ),
+      "no parameter named 'x'"
+  end
+
+  it "says no parameters named" do
+    assert_error %(
+      def foo
+      end
+
+      foo(x: 1, y: 2)
+      ),
+      "no parameters named 'x', 'y'"
+  end
+
+  it "says argument already specified" do
+    assert_error %(
+      def foo(x)
+      end
+
+      foo(1, x: 2)
+      ),
+      "argument for parameter 'x' already specified"
+  end
+
+  it "says type mismatch for positional argument" do
+    assert_error %(
+      def foo(x : Int32, y : Int32)
+      end
+
+      foo(1, 'a')
+      ),
+      "expected argument #2 to 'foo' to be Int32, not Char"
+  end
+
+  it "says type mismatch for positional argument with two options" do
+    assert_error %(
+      def foo(x : Int32)
+      end
+
+      def foo(x : String)
+      end
+
+      foo('a')
+      ),
+      "expected argument #1 to 'foo' to be Int32 or String, not Char"
+  end
+
+  it "says type mismatch for positional argument with three options" do
+    assert_error %(
+      def foo(x : Int32)
+      end
+
+      def foo(x : String)
+      end
+
+      def foo(x : Bool)
+      end
+
+      foo('a')
+      ),
+      "expected argument #1 to 'foo' to be Bool, Int32 or String, not Char"
+  end
+
+  it "says type mismatch for named argument " do
+    assert_error %(
+      def foo(x : Int32, y : Int32)
+      end
+
+      foo(y: 1, x: 'a')
+      ),
+      "expected argument 'x' to 'foo' to be Int32, not Char"
+  end
+
+  it "replaces free variables in positional argument" do
+    assert_error %(
+      def foo(x : T, y : T) forall T
+      end
+
+      foo(1, 'a')
+      ),
+      "expected argument #2 to 'foo' to be Int32, not Char"
```

Comment:
It seems a bit odd that the type restriction for the second argument is fixed to `Int32`, just because the first argument has that type. The actual signature doesn't require `Int32`, it just needs both arguments to have the same type.
To demonstrate that, the inverted error "expected first argument to 'foo' to be Char, not Int32" would be equally correct.
So I think the previous message "no overload matches 'foo' with types Int32, Char" is actually better here for this specific method (it might not be that good when there are more arguments and some of them match).
Alternatively, we could consider improving the message for the case of a free variable bound to multiple arguments to something like "expected first and second argument to 'foo' to be of the same type".

Reply:
Unfortunately, this is exactly how free variable matching works in the compiler right now, so `foo(1 || 'a', 1)` works and `foo(1, 1 || 'a')` doesn't. A somewhat opposite wording can be observed if one calls `foo(y: 'a', x: 1)`.

Reply:
I agree about this improvement, and I also was sure that this was going to be a controversial topic. However, I have no idea how to detect this generally. And I also don't think this case is very common. And we are still providing all of the overloads.

Also, given that T is inferred from the first argument type, it might make sense that the error says that the second argument doesn't match.